language/move-vm/types/src/native_functions/lcs.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    loaded_data::types::Type,
+    native_functions::dispatch::{native_gas, NativeResult},
+    values::{values_impl::Reference, Value},
+};
+use libra_types::vm_error::{StatusCode, VMStatus};
+use std::collections::VecDeque;
+use vm::{
+    errors::VMResult,
+    gas_schedule::{CostTable, NativeCostIndex},
+};
+
+/// Rust implementation of Move's `native public fun serialize<T>(&T): vector<u8>`
+pub fn native_serialize(
+    mut ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+    cost_table: &CostTable,
+) -> VMResult<NativeResult> {
+    if ty_args.len() != 1 {
+        let msg = format!(
+            "Wrong number of type arguments for serialize. Expected 1, but found {}",
+            ty_args.len()
+        );
+        return Err(VMStatus::new(StatusCode::UNREACHABLE).with_message(msg));
+    }
+    if args.len() != 1 {
+        let msg = format!(
+            "Wrong number of arguments for serialize. Expected 1, but found {}",
+            args.len()
+        );
+        return Err(VMStatus::new(StatusCode::UNREACHABLE).with_message(msg));
+    }
+
+    let arg_type = ty_args.pop().unwrap();
+    // delegate to the LCS serialization for `Value`
+    // TODO: add sub status for failed serialization
+    let ref_to_val = pop_arg!(args, Reference);
+    let serialized_value = ref_to_val
+        .read_ref()?
+        .simple_serialize(&arg_type)
+        .ok_or(VMStatus::new(StatusCode::NATIVE_FUNCTION_ERROR))?;
+    // cost is proportional to the size of the serialized value
```

Comment:
cc @tzakian for whether this is reasonable

Reply:
LGTM

Reply:
@sblackshear @tzakian Do you have a unit type in Move? You may need to forbid `vec![(); 10000000]` and such

Reply:
No unit type, and no macro like `vec![(); 10000]`